File: app/scripts/import-to-firestore.js
#!/usr/bin/env node
// Imports exported RTDB JSON into Firestore (mhtt-tournament-a3e15)
// Usage: node app/scripts/import-to-firestore.js

import { initializeApp, applicationDefault } from 'firebase-admin/app';
import { getFirestore } from 'firebase-admin/firestore';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const data = JSON.parse(readFileSync(join(__dirname, 'import-export.json'), 'utf8'));

initializeApp({ credential: applicationDefault(), projectId: 'mhtt-tournament-a3e15' });
const db = getFirestore();

const BATCH_SIZE = 400;

async function importCollection(collectionName, records, idFn) {
  const items = Object.values(records);
  console.log(`Importing ${items.length} ${collectionName}...`);
  for (let i = 0; i < items.length; i += BATCH_SIZE) {
    const batch = db.batch();
    items.slice(i, i + BATCH_SIZE).forEach(item => {
      batch.set(db.collection(collectionName).doc(idFn(item)), item);
    });
    await batch.commit();
  }
  console.log(`✓ ${items.length} ${collectionName} imported`);
}

async function run() {
  if (data.players)    await importCollection('players',     data.players,     p => p.id);
  if (data.rankings)   await importCollection('rankings',    data.rankings,    r => r.name.replace(/[.#$[\]/]/g, '_'));
  if (data.tournaments) await importCollection('tournaments', data.tournaments, t => t.id);
  console.log('\nDone! Check Firestore console to verify.');
}

run().catch(err => { console.error(err); process.exit(1); });

File: backend/import_to_firestore.py
#!/usr/bin/env python3
import json, os, re, sys
import firebase_admin
from firebase_admin import credentials, firestore

DATA_FILE = os.path.join(os.path.dirname(__file__), '../app/scripts/import-export.json')

def init():
    sa_json = os.getenv('SERVICE_ACCOUNT_JSON')
    if sa_json:
        cred = credentials.Certificate(json.loads(sa_json))
        firebase_admin.initialize_app(cred)
    else:
        firebase_admin.initialize_app(options={'projectId': 'mhtt-tournament-a3e15'})
    return firestore.client()

def safe_id(name):
    return re.sub(r'[.#$\[\]/]', '_', name)

def import_collection(db, collection, records, id_fn):
    items = list(records.values())
    print(f'Importing {len(items)} {collection}...')
    batch = db.batch()
    count = 0
    for item in items:
        batch.set(db.collection(collection).document(id_fn(item)), item)
        count += 1
        if count % 400 == 0:
            batch.commit()
            batch = db.batch()
    if count % 400 != 0:
        batch.commit()
    print(f'✓ {len(items)} {collection} imported')

def main():
    with open(DATA_FILE) as f:
        data = json.load(f)
    db = init()
    if 'players' in data:
        import_collection(db, 'players', data['players'], lambda p: p['id'])
    if 'rankings' in data:
        import_collection(db, 'rankings', data['rankings'], lambda r: safe_id(r['name']))
    if 'tournaments' in data:
        import_collection(db, 'tournaments', data['tournaments'], lambda t: t['id'])
    print('\nDone! Verify in Firestore console.')

main()
